msteams/provider/client.py

```diff
@@ -0,0 +1,253 @@
+            self._prepare_attachment_download_url(attachment)
+            for attachment in attachments
+        ]
+        return await asyncio.gather(*downloadable_attachments)
+
+    async def _prepare_attachment_download_url(self, attachment):
+        # prepare url to request download link see
+        # https://learn.microsoft.com/en-us/answers/questions/1072289/download-directly-chat-attachment-using-contenturl
+        content_url = attachment["contentUrl"]
+        b64encoded = base64.b64encode(content_url.encode("utf-8")).decode("utf-8")
+        prepared = "u!" + b64encoded.rstrip("=").replace("/", "_").replace("+", "-")
+        graph_api_url = (
+            f"https://graph.microsoft.com/v1.0/shares/{prepared}/driveItem/content"
+        )
+        async with self.session.get(
+            graph_api_url, headers={"Authorization": f"Bearer {self.access_token}"}
+        ) as response:
+            content = await response.content.read()
+            if not response.ok:
+                return attachment
+            attachment["content"] = content
+            return attachment
+
+    def _prepare_attachments(self, messages, results):
+        attachments = []
+        for message in messages:
+            if message is not None:
+                if "attachments" in message:
+                    for attachment in message["attachments"]:
+                        attachment["downloadUrl"] = None
+                        if attachment["contentType"] == "reference":
+                            if "contentUrl" in attachment:
+                                attachments.append(attachment)
+                results.append(message)
+        if len(attachments) > 0:
+            self.loop.run_until_complete(
+                self._gather_downloadable_attachments(attachments)
+            )
+        # Close session and loop
+        self.loop.run_until_complete(self._close_session())
+        self._close_loop()
+        return attachments
+
+    def _process_hits(self, hits):
+        results = []
+        messages = self.loop.run_until_complete(self._gather_messages(hits))
+        self._prepare_attachments(messages, results)
+        return results
+
+    def _delegated_search(self, query):
+        results = []
+        response = requests.post(
+            self.SEARCH_URL,
+            headers={"Authorization": f"Bearer {self.access_token}"},
+            json={
+                "requests": [
+                    {
+                        "entityTypes": ["chatMessage"],
+                        "query": {
+                            "queryString": query,
+                            "size": self.search_limit,
+                        },
+                    }
+                ]
+            },
+        )
+        if not response.ok:
+            raise UpstreamProviderError(
+                f"Error while searching Outlook: {response.text}"
+            )
+        for hit_container in response.json()["value"][0]["hitsContainers"]:
```

Comment:
I think it's more clear to return the `response.json()["value"][0]["hitsContainers"]` here, and in the `provider.py` do the processing